Add "Last Block" and "Last Tx" rows to peer details area

jonatack opened the pull request:
- add `RPCConsole::TimeDurationField` helper to replace repeated code and call system time only once in `RPCConsole::updateDetailWidget`
- add "Last Tx" (`CNodeStats::nLastTXTime`) field to peer details
- add "Last Block" (`CNodeStats::nLastBlockTime`) field to peer details

jonatack commented:
![Screenshot from 2021-02-24 17-03-13](https://user-images.githubusercontent.com/2415484/109030873-1b5de680-76bc-11eb-901f-b61260634a73.png)
![Screenshot from 2021-02-24 17-02-58](https://user-images.githubusercontent.com/2415484/109030881-1c8f1380-76bc-11eb-8c1d-948bd1816193.png)

jarolrod commented on src/qt/forms/debugwindow.ui:
```diff
@@ -1342,13 +1342,65 @@
                <item row="12" column="0">
+                <widget class="QLabel" name="peerLastBlockLabel">
+                 <property name="toolTip">
+                  <string>Elapsed time since a novel block passing initial validity checks was received from this peer.</string>
```
```suggestion
                  <string>Elapsed time since a novel block was received from this peer.</string>
```

jarolrod commented on src/qt/forms/debugwindow.ui:
```diff
@@ -1342,13 +1342,65 @@
+               <item row="13" column="0">
+                <widget class="QLabel" name="peerLastTxLabel">
+                 <property name="toolTip">
+                  <string>Elapsed time since a novel transaction accepted into our mempool was received from this peer.</string>
```
```suggestion
                  <string>Elapsed time since a novel transaction was received from this peer.</string>
```

jarolrod replied:
alternative
```suggestion
                  <string>Elapsed time since a novel transaction, accepted into our mempool, was received from this peer.</string>
```

jarolrod reviewed (commented):
ACK c2758983f21415adde73f6948009ff3ef7f9bdb2

Small nits, maybe the extra explanation is not needed. Feel free to ignore.

jonatack commented:
Thanks for having a look, @jarolrod! Much appreciated.

The docs echo the ones I added in the `-netinfo help` and are a compact version of those I added to `src/net.h`:
```
    /** UNIX epoch time of the last block received from this peer that we had
     * not yet seen (e.g. not already received from another peer), that passed
     * preliminary validity checks and was saved to disk, even if we don't
     * connect the block or it eventually fails connection. Used as an inbound
     * peer eviction criterium in CConnman::AttemptToEvictConnection. */

    /** UNIX epoch time of the last transaction received from this peer that we
     * had not yet seen (e.g. not already received from another peer) and that
     * was accepted into our mempool. Used as an inbound peer eviction criterium
     * in CConnman::AttemptToEvictConnection. */
```
which I think are a bit more useful to users than the equivalent ones I added in the getpeerinfo rpc help (I've been adding these docs everywhere, hehe). For example, if we receive a novel block that does not pass initial validity checks, or a novel transaction that isn't accepted into our mempool, it won't appear. A user might wonder why, but the answers are difficult for casual users to glean from reading the code, and while writing the docs it took some review discussion with others to get this info cleared up. So it seems worth documenting when we have the space to do so.

jarolrod commented:
@jonatack thanks for the context, looks good to me then 👍

hebasto commented:
Concept ACK. Will review in the morning :)

hebasto commented on src/qt/rpcconsole.cpp:
```diff
@@ -1120,11 +1120,12 @@ void RPCConsole::updateDetailWidget()
-    ui->peerLastSend->setText(stats->nodeStats.nLastSend ? GUIUtil::formatDurationStr(GetSystemTimeInSeconds() - stats->nodeStats.nLastSend) : tr("never"));
-    ui->peerLastRecv->setText(stats->nodeStats.nLastRecv ? GUIUtil::formatDurationStr(GetSystemTimeInSeconds() - stats->nodeStats.nLastRecv) : tr("never"));
+    const int64_t time_now{GetSystemTimeInSeconds()};
+    ui->peerConnTime->setText(TimeDurationField(stats->nodeStats.nTimeConnected, time_now));
```
This refactoring changes the logic of the `peerConnTime` value setting, as now "Never" is a possible option.
Does it makes sense? Is it guaranteed to be safe?

jonatack replied:
Good point. I was seeing output of `0s` and not "Never" but no reason to change the logic. Changed.

jonatack commented:
Updated per `git diff c275898 a5b349` to not use the helper for `peerConnTime` (and inverted the order of the helper arguments).

hebasto reviewed (approved):
ACK a5b349af5878c8127e89abb21d96863a8debbc40, tested on Linux Mint 20.1 (Qt 5.12.8):

![Screenshot from 2021-02-25 13-15-21](https://user-images.githubusercontent.com/32963518/109145848-ba8de700-776b-11eb-9246-c0fb294f2832.png)

![Screenshot from 2021-02-25 13-14-58](https://user-images.githubusercontent.com/32963518/109145861-beba0480-776b-11eb-9e80-bf4321181bac.png)


piconit: while here why not to fix indentation from #179:
```diff
--- a/src/qt/forms/debugwindow.ui
+++ b/src/qt/forms/debugwindow.ui
@@ -1079,7 +1079,7 @@
                <item row="1" column="0">
                 <widget class="QLabel" name="peerConnectionTypeLabel">
                  <property name="toolTip">
-                   <string>The direction and type of peer connection: %1</string>
+                  <string>The direction and type of peer connection: %1</string>
                  </property>
                  <property name="text">
                   <string>Direction/Type</string>
```

jonatack commented:
Thanks @hebasto, done.

hebasto reviewed (approved):
re-ACK 70d3c5d0b987411873d557440607aa4f3f65b1ec

Thank you @jonatack !

jarolrod reviewed (commented):
re-ack 70d3c5d0b987411873d557440607aa4f3f65b1ec
